### ISSUE TITLE
OpenApi spec: self-referencing schema contains ambiguous object description

### ISSUE THREAD

**Issue by andreskask123** [stablekernel/aqueduct#663]
When some ManagedObject class contains self-referencing field (i.e Person class contains "parent" field of type Person) then OpenApi3 specification json, generated by aqueduct cli, describes the field with inline description. Expected behaviour would be reference to component schema object.

Please see: https://gist.github.com/andreskask123/6bcf6b3421a0a2dd897e35ea5d1ab5de

Gist consists of 3 files:
- model.dart
- current_spec.json
- expected_spec.json

Current result: current_spec.json:L350
Expected result is in expected_spec.json:L350.

Note that photo field get's expected descriptions in spec files.

Used cli version: 3.2.1

**Comment by joeconwaystk** [comment]
This is intentional - but willing to evaluate other solutions. The intent was to prevent a user of the API documentation from thinking they could pass a full Person object as `parent`; only the `id` is going to be used. Does this cause issues with code generation or other behavior?

**Comment by andreskask123** [comment]
I see, it does make sense if you put it like that. On the other hand i think of openapi spec as platform agnostic. Which means i would assume it to ignore some specific framework (Aqueduct in this case) behavioural expectations. Let me explain a bit further.
My ideal dev workflow would be that when i write api code i get auto generated sdk for free to simplify api communication and reduce the need for boilerplate code i need to write. Sdk generation would first create all the model classes that’s described in openapi components->schemas and after that it would generate the http communication related code(authorization, error handling, mocking for tests, etc). Those auto-generated model classes i would re-use in my frontend (Angular and Flutter) application codebases. Current code generation solution that i’ve tried (https://github.com/OpenAPITools/openapi-generator) didn’t quite satisfied my needs. Pluss i’d rather have this generation done by dart (just to keep things simple). 
Since the sdk generation isn’t that difficult i started to toy around with the idea myself. Then i bumped into a problem that i can’t quite replicate the model, described in aqueduct api project, since i don’t have enough info on model class properties (i.e Person.parent). In specification it says it’s an object but what other properties does it contain, besides id, is unclear for me. Another example would be that when i pull an object from an api (ie with GET /persons/{id}) i see that it should return Person object with property called parent which is an object with only one property - id. Actually it’s type is Person. Plus if i compare Person.parent description with Parent.photo description then in reality they both refer to a class (Person and Photo respectively) but their openapi components->schema descriptions are totally different (Parent.photo is $ref and Photo.parent is an inline object description). So personally for me it’s unconsistent.
So i did some quick digging in aqueduct’s source and found a really quick (https://github.com/andreskask123/aqueduct/commit/c3fcc34509419434c332a7bebac6a2b8a6e3d06f) solution for openapi spec generation, which works for me but i’m not quite sure how well it fits with the whole Aqueduct’s codebase and if it breaks anything or not. So that’s the reason why i haven’t made a PR yet. 
The solution treats Person.parent and Person.photo the same way and generates similar $ref descriptions both for Person.parent and for Person.photo properties.
In regards of passing the full Person object as a parent, maybe the solution would be to throw an 400 error with correct description? Not fully convinced in the solution myself but rather thinking out loud.

**Comment by andreskask123** [comment]
btw, if it makes sense then i'm willing to share the sdk generation solution as well, when it's mature enough. For me, it's essential part of my (Aqueduct, Angular, Flutter) dev bundle and thanks to static analyzer, it makes my life a lot easier.

**Comment by joeconwaystk** [comment]
One thing to point out is that you can override a component/response body/request body/etc. on a per model/endpoint basis without having to alter the source of the framework. For example, you can override [this method](https://pub.dev/documentation/aqueduct/latest/aqueduct/ManagedObject/documentSchema.html) in your managed object or [this method](https://pub.dev/documentation/aqueduct/latest/aqueduct/ResourceController/documentOperationRequestBody.html) in a resource controller.

In any of these methods, you can call the superclass' implementation and then edit the parts that need to be edited.

**Comment by andreskask123** [comment]
Yes, thanks for pointing it out. Actually i was aware of that solution too, but for me, it seems a little bit like a workaround since i have to implement this method every time i have self reference in some class description. Luckily it doesn't happen often so it's not that big of a problem.

But what are your thoughts on the proposed solution?

**Comment by joeconwaystk** [comment]
I'm having a hard time understanding. If you consider how Aqueduct handles foreign keys using nested objects, i.e.:

```json
{
  "id": 1,
  "name": "Bob",
  "parent" : {
    "id": 2
  }
}
```

Many other frameworks would not use a nested object and instead flatten the structure into the following:

```json
{
  "id": 1,
  "name": "Bob",
  "parent_id" : 2
}
```

When dealing with foreign key fields, most endpoints are only going to expose and accept the foreign key value. That is why this behavior is the default - the endpoint will never accept or expose any other fields of the parent. If you change this behavior so that the endpoint either joins on the parent (and therefore expose more of its fields) or it updates the values of the parent alongside the child, then you need to update your OpenAPI doc to reflect that. That's why the callbacks exist and the default decision is the most common decision.

That being said, I may be misunderstanding you.
